### PR TITLE
Complete rewrite of MHCnames

### PR DESCRIPTION
**Pull request by iskandr**
The previous versions of MHCnames were not well specified with regards to which alleles could be parsed or what the standard output format should look like. Many special exceptions for human and mouse alleles were sprinkled throughout the code to make them work better, but even still there were many alleles in IEDB which could not be parsed. Furthermore, 6 and 8 digit allele resolutions did not work. 

This is a total rewrite of MHCnames which parses strings into any of the following objects:

* `Locus`
* `AlleleGroup`
* `NamedAllele`
* `FourDigitAllele`
* `SixDigitAllele`
* `EightDigitAllele`
* `MutantAllele`
* `AlphaBetaPair`

The parsing is directed by a collection of YAML files scraped from various papers and online resources that specify a multi-species gene ontology, common gene aliases, common species aliases, and serotypes.

### CODE REVIEW

File: mhcnames/named_allele.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import print_function, division, absolute_import
+
+from .allele_group import AlleleGroup
+
+class NamedAllele(AlleleGroup):
+
```
Comment (timodonnell) [inline mhcnames/named_allele.py:31]: A brief indication of what this class represents would be helpful

**Comment by timodonnell** [comment]
This looks awesome! Thanks @iskandr !

**Comment by grayfall** [comment]
`MutantAllele` support seems to be bugged. I've tested it on several examples from IEDB, e.g.

```
In [2]: mhcnames.parse('HLA-A*02:01 A150P mutant')
Out[2]: ---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)          
/home/anaconda/conda/envs/mhc/lib/python3.6/site-packages/IPython/core/formatters.py in __call__(self, obj)
    700                 type_pprinters=self.type_printers,
    701                 deferred_pprinters=self.deferred_printers)
--> 702             printer.pretty(obj)                                           
    703             printer.flush()   
    704             return stream.getvalue()                         
                                                                       
/home/anaconda/conda/envs/mhc/lib/python3.6/site-packages/IPython/lib/pretty.py in pretty(self, obj)
    400                         if cls is not object \
    401                                 and callable(cls.__dict__.get('__repr__')):
--> 402                             return _repr_pprint(obj, self, cycle)     
    403
    404             return _default_pprint(obj, self, cycle)

/home/anaconda/conda/envs/mhc/lib/python3.6/site-packages/IPython/lib/pretty.py in _repr_pprint(obj, p, cycle)
    695     """A pprint that just redirects to the normal repr function."""
    696     # Find newlines and replace them with p.break_()
--> 697     output = repr(obj)
    698     for idx,output_line in enumerate(output.splitlines()):
    699         if idx:

/home/anaconda/conda/envs/mhc/lib/python3.6/site-packages/serializable/serializable.py in __repr__(self)
     37
     38     def __repr__(self):
---> 39         return str(self)
     40
     41     def __eq__(self, other):

/home/anaconda/conda/envs/mhc/lib/python3.6/site-packages/serializable/serializable.py in __str__(self)
     34         return "%s(%s)" % (
     35             self.__class__.__name__,
---> 36             ", ".join("%s=%s" % (k, v) for (k, v) in self.to_dict().items()))
     37
     38     def __repr__(self):

~/storage/melanoma/mhc_binding/mhcnames/mhcnames/mutant_allele.py in to_dict(self)
     40         d = Gene.to_dict(self)
     41         d["original_allele"] = self.original_allele.to_dict()
---> 42         d["original_allele_class"] = self.original_allele_class
     43         d["mutations"] = [m.to_dict() for m in self.mutations]
     44         return d

AttributeError: 'MutantAllele' object has no attribute 'original_allele_class'
```

**Comment by iskandr** [comment]
Closing since this got moved over to MHCgnomes

**Pull request closed without merging**